### PR TITLE
improved obstruction handling in CTS

### PR DESCRIPTION
**Pull request by precisionmoon**
1. move legal & illegal branch points to account for parent movement
2. add occupied location tracking to avoid collision
3. add HTree print routine for debug

### CODE REVIEW

File: src/cts/src/TreeBuilder.h [hunk 1]
```diff
@@ -50,6 +50,29 @@ namespace utl {
 class Logger;
 }  // namespace utl
 
+// Hash and compare functions for occupiedLocations_ map
+// Use this hash table to keep track of occupied locations
+namespace std {
+template <>
+struct hash<cts::Point<double>>
```
Comment (maliberty) [inline src/cts/src/TreeBuilder.h:57]: Could you just use the integer point like odb::Point?

File: src/cts/src/TreeBuilder.h [hunk 2]
```diff
@@ -50,6 +50,29 @@ namespace utl {
...
+{
+  std::size_t operator()(const cts::Point<double>& point) const
+  {
+    return point.getX() + (point.getY() * 11.0);
```
Comment (maliberty) [inline src/cts/src/TreeBuilder.h:61]: This is a weak hash fn - better to use boost::hash for a pair

File: src/cts/src/TreeBuilder.h [hunk 3]
```diff
@@ -120,11 +143,34 @@ class TreeBuilder
...
   }
+  bool isAlongBbox(double x,
+                   double y,
+                   double x1,
+                   double y1,
+                   double x2,
+                   double y2)
+  {
+    if ((floatEqual(x, x1) || (floatEqual(x, x2)))
+        && (floatEqualOrGreater(y, y1) && floatEqualOrSmaller(y, y2))) {
```
Comment (maliberty) [inline src/cts/src/TreeBuilder.h:159]: The main goal of integer DBU is to avoid writing code like this.  Can this be avoided?

**Review by maliberty — COMMENTED** [review]
Initial thoughts before I get into the algorithm.

File: src/cts/test/check_buffers_blockages.ok [hunk 4]
```diff
@@ -46,8 +47,96 @@ TritonCTS forced slew degradation on these wires.
...
+HTree after legalization -------
+HTree: top buf loc:[(7.358, 6.302)]
+HTree: level*0* bufId*0*: branchPt:[(3.571, 6.089)] topo len:4.000 dist to parent:4.000 weighted sink len:74.278 parentPt:[(7.358, 6.302)]
+HTree: level*0* bufId*1*: branchPt:[(11.358, 6.302)] topo len:4.000 dist to parent:4.000 weighted sink len:68.516 parentPt:[(7.358, 6.302)]
+-------------------------------------------------
+HTree: level*1* bufId*0*: branchPt:[(4.868, 5.386)] topo len:2.000 dist to parent:2.000 weighted sink len:30.017 parentPt:[(3.571, 6.089)]
+HTree: level*1* bufId*1*: branchPt:[(3.757, 7.904)] topo len:2.000 dist to parent:2.000 weighted sink len:30.144 parentPt:[(3.571, 6.089)]
+HTree: level*1* bufId*2*: branchPt:[(11.156, 4.505)] topo len:2.000 dist to parent:2.000 weighted sink len:23.951 parentPt:[(11.358, 6.302)]
+HTree: level*1* bufId*3*: branchPt:[(11.861, 7.800)] topo len:2.000 dist to parent:2.000 weighted sink len:26.819 parentPt:[(11.358, 6.302)]
+-------------------------------------------------
```
[the remark this answers is not in the document]
Comment (precisionmoon) [reply src/cts/test/check_buffers_blockages.ok:139]: Correct.  I added debug printing for a couple of regression tests.   They are not visible in default mode.  Is this OK?

File: src/cts/test/array.tcl [hunk 5]
```diff
@@ -130,6 +130,7 @@ set_wire_rc -signal -layer $wire_rc_layer
 set_wire_rc -clock  -layer $wire_rc_layer_clk
 set_dont_use $dont_use
 
+set_debug_level CTS legalizer 3
```
[the remark this answers is not in the document]
Comment (precisionmoon) [reply src/cts/test/array.tcl:133]: I can remove it from array.tcl but I'd like to keep it for one regression test.

File: src/cts/src/Util.h [hunk 6]
```diff
@@ -35,11 +35,20 @@
...
 
 namespace cts {
 
+bool floatEqual(double x1, double x2, double epsilon = 1e-6);
+
+// x1 >= x2
+bool floatEqualOrGreater(double x1, double x2, double epsilon = 1e-6);
+
+// x1 <= x2
+bool floatEqualOrSmaller(double x1, double x2, double epsilon = 1e-6);
```
[the remark this answers is not in the document]
Comment (precisionmoon) [reply src/cts/src/Util.h:50]: OK.  Let me see if I can leverage STA fuzzy routines.

File: src/cts/src/Util.cpp [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+// Without Util.cp, linker complains about duplicate definitions from
+// cts_unittest.cc related to libcts_lib.a
+
+#include "Util.h"
+
+#include <cmath>
+
+namespace cts {
+
+bool floatEqual(double x1, double x2, double epsilon)
```
[the remark this answers is not in the document]
Comment (precisionmoon) [reply src/cts/src/Util.cpp:46]: OK.  Will do.

File: src/cts/src/Util.cpp [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+//
+///////////////////////////////////////////////////////////////////////////////
+
+// It would be good if this can be folded into Util.h
+// Without Util.cp, linker complains about duplicate definitions from
+// cts_unittest.cc related to libcts_lib.a
```
[the remark this answers is not in the document]
Comment (precisionmoon) [reply src/cts/src/Util.cpp:38]: I'll try inline but this won't be needed if we move to STA fuzzy routine.

File: src/cts/src/TreeBuilder.h (hunk 1 above)
Comment (precisionmoon) [reply src/cts/src/TreeBuilder.h:57]: OK.  Let me look into it.

File: src/cts/src/TreeBuilder.h (hunk 2 above)
Comment (precisionmoon) [reply src/cts/src/TreeBuilder.h:61]: OK, let me look into boost::hash.

File: src/cts/src/TreeBuilder.h (hunk 3 above)
Comment (precisionmoon) [reply src/cts/src/TreeBuilder.h:159]: This may be hard because there are many arithmetic operations on these locations.  Let me see what can be done.

**Comment by QuantamHD** [comment]
@precisionmoon Can you confirm that none of the fuzzy routines are used as comparators for anything that requires strict weak ordering. 

That would be std::sort std::map std::set. Fuzzy comparisons aren't allowed in those constructs.

File: src/cts/test/array.tcl (hunk 5 above)
[the remark this answers is not in the document]
Comment (precisionmoon) [reply src/cts/test/array.tcl:133]: removed

File: src/cts/src/Util.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (precisionmoon) [reply src/cts/src/Util.h:50]: Changed to fuzzyEqual, fuzzyEqalOrGreater, ...   Kept it in cts for now.   I may remove this altogether later.

File: src/cts/src/Util.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (precisionmoon) [reply src/cts/src/Util.cpp:38]: Changed to inline.

File: src/cts/src/TreeBuilder.h (hunk 2 above)
Comment (precisionmoon) [reply src/cts/src/TreeBuilder.h:61]: Replaced it with boost::unordered_set.

**Comment by precisionmoon** [comment]
I'm closing this PR since a new one PR #4113 in private branch will be used going forward.

**Pull request closed without merging**